network.go

```diff
@@ -736,6 +736,15 @@ func (n *network) CreateEndpoint(name string, options ...EndpointOption) (Endpoi
 		ep.ipamOptions[netlabel.MacAddress] = ep.iface.mac.String()
 	}
 
+	if ipam.capability.RequiresEndpointInfo {
+		if ep.ipamOptions == nil {
+			ep.ipamOptions = make(map[string]string)
+		}
+		log.Debugf("Endpoint: %s", ep)
+		ep.ipamOptions[netlabel.EndpointName] = name
```

Comment:
I read that comment and it makes sense. The difference though is this is not a core part of the interface but rather an optional capabilitity-based extension.

Without this, it limits the flexibility and use cases around IPAM. For example, one use case for external IPAM systems is to log and maintain the understanding of which application/service/container/host had which IP address when. This can be necessary in reconstructing the IP ownership in a forensic situation, where you are trying to figure out the extent and causes of a breach. To do this, we really need to know who the IP is for when it is allocated and release.

Similarly, in our IPAM solution we provide a single-pane-of-glass management of all IP addresses throughout the network. Seeing an IP address alone and only knowing it belongs to "Docker" doesn't provided enough useful context for the user.

In fact ideally we would know the container ID. If we had that we could just pull all the rest of the meta-data out-of-band.

The next step for the IPAM extensions I am looking at would be to allow the user to pass arbitrary key-value pairs at the time of IP allocation (i.e., end point connect). Right now, this is possible at network creation. But if we do it at IP allocation time, then we can enable the user to control additional meta-data (for example, application name or service name) to be presented in our external system. Similarly, this can enable policy-based addressing; for example, if the user wants to place particular services at particular IP addresses within each subnet (gateways at .1, etc.).

In any case, all of this would be *optional* based on capabilities. I agree they shouldn't be "one big interface" as described in the comments. But by not allowing them as optional capabilities, you are limiting the end user to very narrow use cases.